hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+    always_assert(convert_ok && "Unable to convert instruction"
+                  " to encoder request");
+    xedError = xed_encode(&request, destination, XED_MAX_INSTRUCTION_BYTES,
+                          &encodedSize);
+    always_assert_flog(xedError == XED_ERROR_NONE,
+                       "XED: Error when encoding {}(): {}",
+                       xed_iclass_enum_t2str(instr),
+                       xed_error_enum_t2str(xedError));
+    return encodedSize;
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst0(i, s_xedState, instr, effOperandSizeBits);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr, xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, effOperandSizeBits, dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst1(i, s_xedState, instr,
+                                    effOperandSizeBits, op);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr, const xed_encoder_operand_t& op,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op, effOperandSizeBits, dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst2(i, s_xedState, instr,
+                                    effOperandSizeBits, op_1, op_2);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr,
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, effOperandSizeBits,
+                                   dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   const xed_encoder_operand_t& op_3,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
```

Comment:
We should be able to collapse the `xedEmit` flavors that do and don't take a dest into one function for each number of operands, with something like this:
```c++
uint32_t xedEmit(..., CodeAddress destination = nullptr) {
  auto const size = xedEmitImpl(instr, destination ? destination : dest(), ...);
  if (!destination) codeBlock.moveFrontier(size);
```